Source/DafnyCore/Rewriters/InductionRewriter.cs

```diff
@@ -156,52 +165,149 @@ void ComputeInductionVariables<VarType>(IToken tok, List<VarType> boundVars, Lis
         return;
       }
 
-      // The argument list was legal, so let's use it for the _induction attribute
+      // The argument list was legal, so let's use it for the _induction attribute.
+      // Next, look for matching patterns for the induction hypothesis.
+      if (lemma != null) {
+        var triggers = ComputeAndReportInductionTriggers(lemma, ref attributes, goodArguments, body);
+        if (triggers.Count == 0) {
+          var suppressWarnings = Attributes.Contains(attributes, "nowarn");
+          var warningLevel = suppressWarnings ? ErrorLevel.Info : ErrorLevel.Warning;
+
+          Reporter.Message(MessageSource.Rewriter, warningLevel, null, tok,
+            $"Could not find a trigger for the induction hypothesis. Without a trigger, this may cause brittle verification. " +
+            $"Change or remove the {{:induction}} attribute to generate a different induction hypothesis, or add {{:nowarn}} to silence this warning. " +
+            $"For more information, see the section quantifier instantiation rules in the reference manual.");
+        }
+      }
+
       attributes = new Attributes("_induction", goodArguments, attributes);
+
       return;
     }
 
     // Okay, here we go, coming up with good induction setting for the given situation
     var inductionVariables = new List<Expression>();
-    if (lemma != null && !lemma.IsStatic) {
-      if (args != null || searchExprs.Exists(expr => FreeVariablesUtil.ContainsFreeVariable(expr, true, null))) {
+    if (lemma is { IsStatic: false }) {
+      if (args != null || InductionHeuristic.VarOccursInArgumentToRecursiveFunction(Reporter.Options, body, null)) {
         inductionVariables.Add(new ThisExpr(lemma));
       }
     }
 
     foreach (IVariable n in boundVars) {
-      if (!(n.Type.IsTypeParameter || n.Type.IsAbstractType || n.Type.IsInternalTypeSynonym) && (args != null ||
-            searchExprs.Exists(expr => InductionHeuristic.VarOccursInArgumentToRecursiveFunction(Reporter.Options, expr, n)))) {
+      if (!(n.Type.IsTypeParameter || n.Type.IsAbstractType || n.Type.IsInternalTypeSynonym || n.Type.IsArrowType) &&
+          (args != null || InductionHeuristic.VarOccursInArgumentToRecursiveFunction(Reporter.Options, body, n))) {
         inductionVariables.Add(new IdentifierExpr(n.Tok, n));
       }
     }
 
     if (inductionVariables.Count != 0) {
+      List<List<Expression>> triggers = null;
+      if (lemma != null) {
+        triggers = ComputeInductionTriggers(inductionVariables, body, lemma.EnclosingClass.EnclosingModuleDefinition);
```

Comment:
I find the logic here a bit hard to follow: we have ComputeInductionTriggers and ComputeAndReportInductionTriggers and ReportInductionTriggers, and two sets of calls for each of these methods.  Is the duplication necessary?  Can we add a comment why?

Reply:
I added a comment. (The reason is to get the tooltips to come out in a nice order.)